src/session.rs

```diff
@@ -607,45 +625,70 @@ where
         connector: C,
-    ) -> Result<Client, error::NewSessionError> {
+    ) -> Result<
+        (
+            hyper_util::client::legacy::Client<C, BoxBody<hyper::body::Bytes, Infallible>>,
+            url::Url,
+        ),
+        error::NewSessionError,
+    > {
         // Where is the WebDriver server?
-        let wdb = webdriver.parse::<url::Url>();
-        let wdb = wdb.map_err(error::NewSessionError::BadWebdriverUrl)?;
+        let wdb = webdriver
+            .parse::<url::Url>()
+            .map_err(error::NewSessionError::BadWebdriverUrl)?;
+
         // We want a tls-enabled client
         let client = hyper_util::client::legacy::Client::builder(TokioExecutor::new())
             .build::<_, BoxBody<hyper::body::Bytes, Infallible>>(connector);
 
-        let mut cap = cap.to_owned();
+        Ok((client, wdb))
+    }
+
+    pub(crate) async fn setup_session(
+        client: hyper_util::client::legacy::Client<C, BoxBody<hyper::body::Bytes, Infallible>>,
+        wdb: url::Url,
+        session_id: Option<&str>,
+        cap: Option<webdriver::capabilities::Capabilities>,
+    ) -> Result<Client, error::NewSessionError> {
         // We're going to need a channel for sending requests to the WebDriver host
         let (tx, rx) = mpsc::unbounded_channel();
 
-        // Set up our WebDriver session.
-        tokio::spawn(Session {
-            rx,
-            ongoing: Ongoing::None,
-            client,
-            wdb,
-            session: None,
-            is_legacy: false,
-            ua: None,
-            persist: false,
-        });
+        if let Some(session_id) = session_id {
+            // Reconnect to an existing session
+            tokio::spawn(Session::new(rx, client, wdb, session_id));
+        } else if let Some(_capabilities) = cap {
+            // Spawn a new WebDriver session.
+            tokio::spawn(Session {
+                rx,
+                ongoing: Ongoing::None,
+                client,
+                wdb,
+                session: None,
+                is_legacy: false,
+                ua: None,
+                persist: false,
+            });
+        }
 
         // now that the session is running, let's do the handshake
-        let client = Client {
+        Ok(Client {
             tx: tx.clone(),
             is_legacy: false,
             new_session_response: None,
-        };
+        })
+    }
+
+    pub(crate) async fn with_capabilities_and_connector(
+        webdriver: &str,
+        cap: &webdriver::capabilities::Capabilities,
+        connector: C,
+    ) -> Result<Client, error::NewSessionError> {
+        let (client, wdb) = Self::create_client_and_parse_url(webdriver, connector).await?;
+        let mut cap = cap.to_owned();
 
-        // Create a new session for this client
-        // https://www.w3.org/TR/webdriver/#dfn-new-session
-        // https://www.w3.org/TR/webdriver/#capabilities
-        //  - we want the browser to wait for the page to load
```

Comment:
I think this last line of the comment should stay here as it's specifically related to setting `pageLoadStrategy`.